crates/ruff_python_parser/src/string.rs

```diff
@@ -535,68 +227,95 @@ impl<'a> StringParser<'a> {
+        if self.kind.is_any_bytes() {
+            self.parse_bytes()
         } else {
-            self.parse_string().map(|expr| vec![expr])
+            self.parse_string()
         }
     }
 }
 
-fn parse_fstring_expr(source: &str, location: TextSize) -> Result<Expr, ParseError> {
-    let fstring_body = format!("({source})");
-    parse_expression_starts_at(&fstring_body, "<fstring>", location)
+pub(crate) fn parse_string_literal(
+    start: TextSize,
+    (source, kind, triple_quoted): (String, StringKind, bool),
+) -> Result<Expr, LexicalError> {
+    let start = start
+        + kind.prefix_len()
+        + if triple_quoted {
+            TextSize::from(3)
+        } else {
+            TextSize::from(1)
+        };
+    StringParser::new(source.as_str(), kind, start).parse()
 }
 
-fn parse_string(
-    source: &str,
-    kind: StringKind,
-    triple_quoted: bool,
+pub(crate) fn parse_fstring_middle(
     start: TextSize,
-) -> Result<Vec<Expr>, LexicalError> {
-    StringParser::new(source, kind, triple_quoted, start).parse()
+    (source, is_raw): (String, bool),
+) -> Result<Option<Expr>, LexicalError> {
+    // This is to account for the empty `FStringMiddle` token that is created
+    // to check for non-parenthesized lambda expressions.
+    if source.is_empty() {
+        return Ok(None);
+    }
+    let kind = if is_raw {
+        StringKind::RawString
+    } else {
+        StringKind::String
+    };
+    StringParser::new(source.as_str(), kind, start)
+        .parse_fstring_middle()
+        .map(Some)
 }
 
-pub(crate) fn parse_strings(
-    values: Vec<(TextSize, (String, StringKind, bool), TextSize)>,
-) -> Result<Expr, LexicalError> {
-    // Preserve the initial location and kind.
-    let initial_start = values[0].0;
-    let last_end = values.last().unwrap().2;
-    let initial_kind = (values[0].1 .1 == StringKind::Unicode).then(|| "u".to_owned());
-    let has_fstring = values
+/// Concatenate a list of string expressions into a single string expression.
+///
+/// This is mainly used for implicit string concatenation and the possible
+/// expression values are strings, bytes, and f-strings.
+pub(crate) fn concatenate_strings(
+    strings: Vec<ParenthesizedExpr>,
+    range: TextRange,
+) -> Result<ParenthesizedExpr, LexicalError> {
+    #[cfg(debug_assertions)]
+    debug_assert!(!strings.is_empty());
+
+    let has_fstring = strings
         .iter()
-        .any(|(_, (_, kind, ..), _)| kind.is_any_fstring());
-    let num_bytes = values
+        .any(|parenthesized_expr| parenthesized_expr.expr.is_f_string_expr());
+    let num_bytes = strings
         .iter()
-        .filter(|(_, (_, kind, ..), _)| kind.is_any_bytes())
+        .filter(|parenthesized_expr| {
+            matches!(
+                parenthesized_expr.expr,
+                Expr::Constant(ast::ExprConstant {
+                    value: Constant::Bytes(_),
+                    ..
+                })
+            )
+        })
         .count();
```

Comment:
Nit: this requires two full iterations. We can reduce it to one by using a plain loop

```rust
let mut has_fstring = false;
let mut byte_literal_count = 0u32;

for string in strings {
    has_fstring |= pexpr.is_f_string_expr();

    if is_bytes {
        byte_literal_count += 1;
    }
}
```